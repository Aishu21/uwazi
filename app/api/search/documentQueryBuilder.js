
export default function () {
  let baseQuery = {
    _source: {
      include: [ 'doc.title', 'doc.processed', 'doc.creationDate', 'doc.template', 'doc.metadata', 'doc.type']
    },
    from: 0,
    size: 12,
    query: {
      bool: {
        must: [
          {match: {'doc.published': true}}
        ]
      }
    },
    filter: {
      bool: {
        must: []
      }
    },
    sort: [],
    aggregations: {
      types: {
        terms: {
          field: 'doc.template.raw',
          size: 0
        },
        aggregations: {
          filtered: {
            filter: {
              bool: {
                must: []
              }
            }
          }
        }
      }
    }
  };

  return {
    query() {
      return baseQuery;
    },

    fullTextSearch(term, fieldsToSearch = ['doc.fullText', 'doc.title']) {
      if (term) {
        baseQuery.query.bool.must.push({
          multi_match: {
            query: term,
            type: 'phrase_prefix',
            fields: fieldsToSearch
          }
        });
      }
      return this;
    },

    sort(property, order = 'desc') {
      let sort = {};
      sort[`doc.${property}`] = {order, ignore_unmapped: true};
      baseQuery.sort.push(sort);
      return this;
    },

    textFilter(filters, property) {
      let match = {match: {}};
      match.match[`doc.metadata.${property}`] = filters[property].value;
      return match;
    },

    rangeFilter(filters, property) {
      let match = {range: {}};
      match.range[`doc.metadata.${property}`] = {gte: filters[property].value.from, lte: filters[property].value.to};
      return match;
    },

    multiselectFilter(filters, property) {
      let values = filters[property].value;
      let match = {terms: {}};
      match.terms[`doc.metadata.${property}.raw`] = values;
      return match;
    },

    nestedFilter(filters, property) {
      let match = {
        nested: {
          path: `doc.metadata.${property}`,
          filter: {
            bool: {}
          }
        }
      };
      let value = filters[property].value;
      let strictSearch = value.strict;
      let properties = value.properties || [];

      let keys = Object.keys(properties).filter((key) => {
        return properties[key].any ||
               properties[key].values;
      });

      if (strictSearch) {
        match.nested.filter.bool.must = [];
        keys.forEach((key) => {
          if (properties[key].any) {
            match.nested.filter.bool.must.push({exists: {field: `doc.metadata.${property}.${key}`}});
            return;
          }

          properties[key].values.forEach((val) => {
            let term = {term: {}};
            term.term[`doc.metadata.${property}.${key}`] = {value: val};
            match.nested.filter.bool.must.push(term);
          });
        });
      }

      if (!strictSearch) {
        match.nested.filter.bool.should = keys.map((key) => {
          if (properties[key].any) {
            return {exists: {field: `doc.metadata.${property}.${key}`}};
          }

          let terms = {terms: {}};
          terms.terms[`doc.metadata.${property}.${key}`] = properties[key].values;
          return terms;
        });
      }

      return match;
    },

    filterMetadata(filters = {}) {
      Object.keys(filters).forEach((property) => {
        let match;
        if (filters[property].type === 'text') {
          match = this.textFilter(filters, property);
        }

        if (filters[property].type === 'range') {
          match = this.rangeFilter(filters, property);
        }

        if (filters[property].type === 'multiselect') {
          match = this.multiselectFilter(filters, property);
        }

        if (filters[property].type === 'nested') {
          match = this.nestedFilter(filters, property);
        }

        baseQuery.filter.bool.must.push(match);
        baseQuery.aggregations.types.aggregations.filtered.filter.bool.must.push(match);
      });
      return this;
    },

    aggregation(key, filters) {
      return {
        terms: {
          field: key,
          size: 0
        },
        aggregations: {
          filtered: {
            filter: {
              bool: {
                must: filters
              }
            }
          }
        }
      };
    },

    nestedAggregation(property, filters) {
      let nestedAggregation = baseQuery.aggregations[property.name] = {
        nested: {
          path: `doc.metadata.${property.name}`
        },
        aggregations: {}
      };

      property.nestedProperties.forEach((prop) => {
        let nestedFilters = baseQuery.filter.bool.must.filter((match) => {
          return match.nested && !match.nested.filter.bool.should;
        })
        .map((nestedFilter) => nestedFilter.nested.filter.bool.must)
        .reduce((result, propFilters) => {
          return result.concat(propFilters);
        }, []);

        nestedAggregation.aggregations[prop.key] = {
          terms: {
            field: `doc.metadata.${property.name}.${prop.key}.raw`,
            size: 0
          },
          aggregations: {
            filtered: {
              filter: {
                bool: {
                  must: nestedFilters
                }
              },
              aggregations: {
                total: {
                  reverse_nested: {},
                  aggregations: {
                    filtered: {
                      filter: {
                        bool: {
                          must: filters
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        };
      });

      return nestedAggregation;
    },

    aggregations(properties) {
      properties.forEach((property) => {
        let path = `doc.metadata.${property.name}.raw`;
        let filters = baseQuery.filter.bool.must.filter((match) => {
          return !match.terms ||
                 match.terms && !match.terms[path];
        });

        if (property.nested) {
          return baseQuery.aggregations[property.name] = this.nestedAggregation(property, filters);
        }

        baseQuery.aggregations[property.name] = this.aggregation(path, filters);
      });
      return this;
    },

    filterByTemplate(templates = []) {
      if (templates.length) {
        let match = {terms: {'doc.template.raw': templates}};
        baseQuery.filter.bool.must.push(match);
      }
      return this;
    },

    highlight(fields) {
      baseQuery.highlight = {
        pre_tags : ['<b>'],
        post_tags : ['</b>']
      };
      baseQuery.highlight.fields = {};
      fields.forEach((field) => {
        baseQuery.highlight.fields[field] = {};
      });
      return this;
    },

    from(from) {
      baseQuery.from = from;
      return this;
    },

    limit(size) {
      baseQuery.size = size;
      return this;
    }
  };
}
