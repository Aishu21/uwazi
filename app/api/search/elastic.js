import elasticSearch from '@elastic/elasticsearch';

const elastic = new elasticSearch.Client({
  node: process.env.ELASTICSEARCH_URL || 'http://localhost:9200',
});

export default elastic;
