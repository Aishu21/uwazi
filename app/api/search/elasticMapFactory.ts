import { template } from 'api/relationships/specs/fixtures';
import { MetadataTemplate } from 'app/Templates/components/MetadataTemplate';
import { TemplateSchema } from 'shared/types/templateType';
import { isArrayLiteralExpression } from 'typescript';

const text = {
  type: 'text',
  index: true,
  analyzer: 'tokenizer',
  fields: {
    raw: { type: 'keyword' },
    sort: { type: 'text', fielddata: true, analyzer: 'string_sorter' },
  },
  term_vector: 'with_positions_offsets',
};

const noSorttext = {
  type: 'text',
  index: true,
  analyzer: 'tokenizer',
  term_vector: 'with_positions_offsets',
};

const noIndexText = {
  type: 'text',
  index: false,
};

const id = {
  type: 'keyword',
  index: true,
};

const date = { type: 'date', doc_values: true };

const geolocation = { type: 'object' };
const nested = { type: 'nested' };
const number = {
  type: 'double',
  doc_values: true,
  fields: {
    raw: { type: 'double', index: false },
    sort: { type: 'double' },
  },
};

const noSortNumber = {
  type: 'double',
  doc_values: true,
};

const alias = (path, key) => {
  return {
    type: 'alias',
    path: `${path}.${key}`,
  };
};

const textType = () => {
  return {
    label: text,
    value: text,
  };
};

const dateType = () => {
  return {
    label: text,
    value: date,
  };
};

const daterangeType = () => {
  return {
    label: text,
    value: {
      properties: {
        from: date,
        to: date,
      },
    },
  };
};

const geolocationType = () => {
  return {
    value: {
      properties: {
        label: text,
        lat: noSortNumber,
        lon: noSortNumber,
      },
    },
  };
};

const imageType = () => {
  return {
    value: noIndexText,
  };
};

const linkType = () => {
  return {
    value: {
      properties: {
        label: text,
        url: noIndexText,
      },
    },
  };
};

const markdownType = () => {
  return {
    value: noSorttext,
  };
};

const selectType = () => {
  return {
    label: text,
    value: id,
  };
};

const noMapping = () => {
  return;
};

const numericType = () => {
  return {
    value: number,
  };
};

const relationshipType = () => {
  return {
    label: text,
    value: id,
    type: noIndexText,
  };
};

const mappings = {
  text: textType,
  date: dateType,
  daterange: daterangeType,
  geolocation: geolocationType,
  image: imageType,
  link: linkType,
  markdown: markdownType,
  media: imageType,
  multidate: dateType,
  multidaterange: daterangeType,
  multiselect: selectType,
  nested: noMapping,
  numeric: numericType,
  preview: noMapping,
  relationship: relationshipType,
  select: selectType,
};

export default {
  mapping: (templates: TemplateSchema[]) => {
    const baseMappingObject = {
      properties: {
        metadata: {
          properties: {},
        },
      },
    };

    return templates.reduce((baseMapping: any, template: TemplateSchema) => {
      return template.properties?.reduce((map: any, property) => {
        if (!property.name || !property.type) {
          return map;
        }

        map.properties.metadata.properties[`${property.name}_${property.type}`] = {
          properties: mappings[property.type](),
        };

        return map;
      }, baseMapping);
    }, baseMappingObject);
  },

  aliasses: (templates: TemplateSchema[]) => {
    const baseAliasObject = {
      properties: {
        metadata: {
          properties: {},
        },
      },
    };

    return templates.reduce((baseAlias: any, template: TemplateSchema) => {
      return template.properties?.reduce((aliases: any, property) => {
        if (!property.name || !property.type) {
          return aliases;
        }
        const mapping = mappings[property.type]();
        const properties = Object.keys(mapping || {}).reduce((map: any, key) => {
          map[key] = {
            type: 'alias',
            path: `metadata.${property.name}_${property.type}.${key}`,
          };

          return map;
        }, {});

        aliases.properties.metadata.properties[`${property.name}`] = { properties };

        return aliases;
      }, baseAlias);
    }, baseAliasObject);
  },

  ingest: (templates: TemplateSchema[]) => {
    const basePipelineObject = {
      description: 'rename pipeline',
      processors: [],
    };
    return templates.reduce((basePipeline, template) => {
      return template.properties?.reduce((pipeline: any, property) => {
        if (!property.name || !property.type) {
          return pipeline;
        }

        pipeline.processors.push({
          rename: {
            field: `metadata.${property.name}`,
            target_field: `metadata.${property.name}_${property.type}`,
          },
        });

        return pipeline;
      }, basePipeline);
    }, basePipelineObject);
  },
};
