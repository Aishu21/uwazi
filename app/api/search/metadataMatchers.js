const textFilter = (filter) => {
  const match = { match: {} };
  match.match[`metadata.${filter.name}`] = filter.value;
  return match;
};

const rangeFilter = (filter) => {
  const match = { range: {} };
  match.range[`metadata.${filter.name}`] = { gte: filter.value.from, lte: filter.value.to };
  return match;
};

const multiselectFilter = (filter) => {
  const filterValue = filter.value;
  const { values } = filterValue;
  let match;
  if (values.includes('missing') && !filterValue.and) {
    const _values = values.filter(v => v !== 'missing');
    match = {
      bool: {
        should: [
          {
            bool: {
              must_not: [
                {
                  exists: {
                    field: `metadata.${filter.name}.raw`
                  }
                }
              ]
            }
          },
          {
            terms: {
            }
          }
        ]
      }
    };
    match.bool.should[1].terms[`metadata.${filter.name}.raw`] = _values;
    return match;
  }
  if (!values.includes('missing') && !filterValue.and) {
    match = { terms: {} };
    match.terms[`metadata.${filter.name}.raw`] = values;
  }

  if (filterValue.and) {
    match = { bool: { must: [] } };
    match.bool.must = values.map((value) => {
      const m = { term: {} };
      m.term[`metadata.${filter.name}.raw`] = value;
      return m;
    });
  }
  return match;
};

const nestedrangeFilter = (filter) => {
  const match = {
    nested: {
      path: `metadata.${filter.name}`,
      query: {
        bool: {
          should: []
        }
      }
    }
  };
  const fromMatch = { range: {} };
  fromMatch.range[`metadata.${filter.name}.from`] = { gte: filter.value.from, lte: filter.value.to };
  const toMatch = { range: {} };
  toMatch.range[`metadata.${filter.name}.to`] = { gte: filter.value.from, lte: filter.value.to };

  match.nested.query.bool.should.push(fromMatch);
  match.nested.query.bool.should.push(toMatch);
  return match;
};

const strictNestedFilter = (filter) => {
  const match = {
    nested: {
      path: `metadata.${filter.name}`,
      query: {
        bool: {
          must: []
        }
      }
    }
  };

  const { properties } = filter.value;
  if (!properties) {
    return;
  }
  const keys = Object.keys(properties).filter(key => properties[key].any ||
      properties[key].values);

  keys.forEach((key) => {
    if (properties[key].any) {
      match.nested.query.bool.must.push({ exists: { field: `metadata.${filter.name}.${key}` } });
      return;
    }

    properties[key].values.forEach((val) => {
      const term = { term: {} };
      term.term[`metadata.${filter.name}.${key}.raw`] = { value: val };
      match.nested.query.bool.must.push(term);
    });
  });

  return match;
};

const nestedFilter = (filter) => {
  const match = {
    bool: {
      must: []
    }
  };

  const { properties } = filter.value;

  const keys = Object.keys(properties).filter(key => properties[key].any ||
      properties[key].values && properties[key].values.length);
  const nestedMatchers = keys.map((key) => {
    let nestedmatch;
    if (properties[key].any) {
      nestedmatch = {
        nested: {
          path: `metadata.${filter.name}`,
          query: {
            bool: {
              must: [
              ]
            }
          }
        }
      };

      nestedmatch.nested.query.bool.must[0] = { exists: { field: `metadata.${filter.name}.${key}` } };
      return nestedmatch;
    }

    const matchers = properties[key].values
    .filter(val => val !== 'missing')
    .map((val) => {
      const _match = {
        nested: {
          path: `metadata.${filter.name}`,
          query: {
            bool: {
              must: [
                { term: {} }
              ]
            }
          }
        }
      };
      _match.nested.query.bool.must[0].term[`metadata.${filter.name}.${key}.raw`] = val;
      return _match;
    });

    if (properties[key].values.includes('missing')) {
      matchers.push({
        nested: {
          path: `metadata.${filter.name}`,
          query: {
            bool: {
              must_not: [
                {
                  exists: {
                    field: `metadata.${filter.name}.${key}.raw`
                  }
                }
              ]
            }
          }
        }
      });
    }

    return matchers;
  });

  match.bool.must = nestedMatchers.reduce((result, matchers) => result.concat(matchers), []);
  return match;
};

const relationshipfilter = (filter) => {
  const match = {
    has_child: {
      type: 'relationship',
      score_mode: 'max',
      inner_hits: {
        _source: false
      },
      query: {
        bool: {
          filter: filter.filters.map((fil) => {
            fil.value = filter.value[fil.name];
            if (fil.value) {
              return filterToMatch(fil);
            }
          }).filter(m => m)
        }
      }
    }
  };

  return match;
};

const filterToMatch = (filter) => {
  let match;
  if (filter.type === 'text') {
    match = textFilter(filter);
  }

  if (filter.type === 'range') {
    match = rangeFilter(filter);
  }

  if (filter.type === 'multiselect' || filter.type === 'relationship') {
    match = multiselectFilter(filter);
  }

  if (filter.type === 'nested' && filter.value.strict) {
    match = strictNestedFilter(filter);
  }

  if (filter.type === 'nested' && !filter.value.strict) {
    match = nestedFilter(filter);
  }

  if (filter.type === 'nestedrange') {
    match = nestedrangeFilter(filter);
  }

  if (filter.type === 'relationshipfilter') {
    match = relationshipfilter(filter);
  }

  return match;
};

export default filterToMatch;

export {
  textFilter,
  rangeFilter,
  multiselectFilter,
  nestedrangeFilter,
  strictNestedFilter,
  nestedFilter,
  relationshipfilter
};
