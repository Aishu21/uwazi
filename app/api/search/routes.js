import Joi from 'joi';
import entities from 'api/entities';
import search from './search';
import { validateRequest } from '../utils';
import needsAuthorization from '../auth/authMiddleware';

export default (app) => {
  app.get('/api/search/count_by_template',
  validateRequest(Joi.object().keys({
    templateId: Joi.string().required()
  }).required(), 'query'),
  (req, res, next) => entities.countByTemplate(req.query.templateId)
  .then(results => res.json(results))
  .catch(next));

  app.get('/api/search',
    validateRequest(Joi.object().keys({
      filters: Joi.string(),
      types: Joi.string(),
      fields: Joi.string(),
      aggregations: Joi.string()
    }), 'query'),
    (req, res) => {
      if (req.query.filters) {
        req.query.filters = JSON.parse(req.query.filters);
      }
      if (req.query.types) {
        req.query.types = JSON.parse(req.query.types);
      }
      if (req.query.fields) {
        req.query.fields = JSON.parse(req.query.fields);
      }
      if (req.query.aggregations) {
        req.query.aggregations = JSON.parse(req.query.aggregations);
      }

      return search.search(req.query, req.language, req.user)
      .then(results => res.json(results))
      .catch(res.error);
    }
  );

  app.get('/api/search_snippets',
  validateRequest(Joi.object().keys({
    searchTerm: Joi.string().allow(''),
    id: Joi.string()
  }), 'query'),
  (req, res, next) => search.searchSnippets(req.query.searchTerm, req.query.id, req.language)
  .then(results => res.json(results))
  .catch(next));

  app.get('/api/search/unpublished', needsAuthorization(['admin', 'editor']), (req, res, next) => {
    search.getUploadsByUser(req.user, req.language)
    .then(response => res.json({ rows: response }))
    .catch(next);
  });
};
