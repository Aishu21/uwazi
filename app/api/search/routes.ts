import { Application, Request, Response, NextFunction } from 'express';
import search from './search';

export default (app: Application) => {
  app.get('/api/search/lookup', (req: Request, res: Response, next: NextFunction) => {
    const { query } = req;
    search
      .autocomplete(query.searchTerm, req.language, JSON.parse(query.templates))
      .then(response => res.json(response))
      .catch(next);
  });

  app.get('/api/search/lookupaggregation', (req: Request, res: Response, next: NextFunction) => {
    const query = JSON.parse(req.query.query);
    search
      .autocompleteAggregations(
        query,
        req.language,
        req.query.property,
        req.query.searchTerm,
        req.user
      )
      .then(response => res.json(response))
      .catch(next);
  });
};
