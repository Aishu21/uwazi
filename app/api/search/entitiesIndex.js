import languagesUtil from 'shared/languages';
import languages from 'shared/languagesList';
import entities from 'api/entities';
import relationships from 'api/relationships/relationships';
import { entityDefaultDocument } from 'shared/entityDefaultDocument';

import elastic from './elastic';

export class IndexError extends Error {}

const handleErrors = async itemsWithErrors => {
  if (itemsWithErrors.length === 0) return;
  const error = new IndexError('ERROR! Failed to index documents.');
  error.errors = itemsWithErrors;
  throw error;
};

function setFullTextSettings(defaultDocument, id, body, doc) {
  const fullText = Object.values(defaultDocument.fullText).join('\f');

  let language;
  if (!defaultDocument.language) {
    language = languagesUtil.detect(fullText);
  }
  if (defaultDocument.language) {
    language = languages(defaultDocument.language);
  }
  const fullTextObject = {
    [`fullText_${language}`]: fullText,
    fullText: { name: 'fullText', parent: id },
  };
  body.push(fullTextObject);
  delete doc.fullText;
}

const bulkIndex = async (docs, _action = 'index', elasticIndex) => {
  const body = [];
  // eslint-disable-next-line max-statements
  docs.forEach(doc => {
    let docBody = Object.assign({ documents: [] }, doc);
    docBody.fullText = 'entity';
    const id = doc._id.toString();
    ['_id', '_rev', 'pdfInfo'].forEach(e => delete docBody[e]);
    const action = {};
    action[_action] = { _index: elasticIndex, _id: id };
    if (_action === 'update') {
      docBody = { doc: docBody };
    }

    const defaultDocument = { ...(entityDefaultDocument(doc.documents, doc.language, 'en') || {}) };

    docBody.documents.forEach(document => {
      delete document.fullText;
    });

    body.push(action);
    body.push(docBody);

    if (defaultDocument.fullText) {
      body.push({
        [_action]: { _index: elasticIndex, _id: `${id}_fullText`, routing: id },
      });
      setFullTextSettings(defaultDocument, id, body, doc);
    }
  });

  const results = await elastic.bulk({ body, requestTimeout: 40000 });
  if (results.items) {
    await handleErrors(results.items.filter(f => f.index.error));
  }

  return results;
};

const newIndexErrorsOrThrow = err => {
  if (!(err instanceof IndexError)) {
    throw err;
  } else {
    return err.errors;
  }
};

const indexEntities = async (
  query,
  select = '',
  limit = 50,
  { batchCallback = () => {}, elasticIndex, searchInstance }
) => {
  const index = async (offset, totalRows, errors = []) => {
    if (offset >= totalRows) {
      return errors.length ? handleErrors(errors) : Promise.resolve();
    }

    const entitiesToIndex = await entities.get(query, '', {
      skip: offset,
      limit,
      documentsFullText: select && select.includes('+fullText'),
    });

    const entitiesToIndexWithRels = await Promise.all(
      entitiesToIndex.map(entity =>
        relationships
          .get({ entity: entity.sharedId })
          .then(relations => ({ ...entity, relationships: relations || [] }))
      )
    );

    let newIndexErrors = [];

    try {
      await searchInstance
        .bulkIndex(entitiesToIndexWithRels, 'index', elasticIndex)
        .then(() => batchCallback(entitiesToIndexWithRels.length, totalRows));
    } catch (err) {
      newIndexErrors = newIndexErrorsOrThrow(err);
    }

    return index(offset + limit, totalRows, errors.concat(newIndexErrors));
  };

  const totalRows = await entities.count(query);
  return index(0, totalRows);
};

export { bulkIndex, indexEntities };
