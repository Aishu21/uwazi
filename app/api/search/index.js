import documentQueryBuilder from './documentQueryBuilder';
import elastic from './elastic';
import routes from './routes';
import search from './search';

console.log(search);
export default routes;

export {
  routes,
  elastic,
  search,
  documentQueryBuilder
};
