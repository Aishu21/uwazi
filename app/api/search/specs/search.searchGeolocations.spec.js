import db from 'api/utils/testing_db';
import { search } from 'api/search';
import instanceElasticTesting from 'api/utils/elastic_testing';

import inheritanceFixtures, { ids } from './fixturesInheritance';

describe('search.searchGeolocations', () => {
  const elasticTesting = instanceElasticTesting('search.geolocation_index_test');
  const user = { _id: 'u1' };

  beforeAll(async () => {
    await db.clearAllAndLoad(inheritanceFixtures);
    await elasticTesting.reindex();
  });

  afterAll((done) => {
    db.disconnect().then(done);
  });

  const cleanResults = results => results.rows.reduce((_memo, row) => {
    const memo = _memo;
    const { sharedId, metadata } = row;
    memo.push({ sharedId, metadata });
    return memo;
  }, []);

  it('should include all geolocation finds, inheriting metadata', async () => {
    const results = await search.searchGeolocations({ order: 'asc', sort: 'sharedId' }, 'en', user);
    expect(cleanResults(results)).toMatchSnapshot();
  });

  it('should allow filtering as in normal search', async () => {
    const results = await search.searchGeolocations({ types: [ids.template3], order: 'asc', sort: 'sharedId' }, 'en', user);
    expect(cleanResults(results)).toMatchSnapshot();
  });

  it('should not fetch unpublished inherited metadata if request is not authenticated', async () => {
    const results = await search.searchGeolocations({ types: [ids.template3], order: 'asc', sort: 'sharedId' }, 'en');
    const cleaned = cleanResults(results);
    const entity = cleaned.find(e => e.sharedId === 'entity_isLinkedToPrivateEntity');
    expect(entity).toBeFalsy();
    expect(results.rows.length).toBe(2);
    expect(results.totalRows).toBe(2);
  });
});
