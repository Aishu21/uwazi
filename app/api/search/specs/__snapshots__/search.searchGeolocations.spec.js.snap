// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search.searchGeolocations should allow filtering as in normal search 1`] = `Array []`;

exports[`search.searchGeolocations should include all geolocation finds, inheriting metadata 1`] = `Array []`;
